common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+
+                    actualEntry.state.get match {
+                        case Cached => {
+                            logger.debug(this, s"Cached read $key")
+                            makeNoteOfCacheHit(key)
+                            actualEntry.value
+                        }
+
+                        case ReadInProgress => {
+                            if (actualEntry.transid == transid) {
+                                logger.debug(this, "Read initiated");
+                                makeNoteOfCacheMiss(key)
+                                listenForReadDone(key, actualEntry, future, promise)
+                                actualEntry.value
+
+                            } else {
+                                logger.debug(this, "Coalesced read")
+                                makeNoteOfCacheHit(key)
+                                actualEntry.value
+                            }
+                        }
+
+                        case WriteInProgress | InvalidateInProgress => {
+                            logger.debug(this, "Reading around a write in progress")
+                            makeNoteOfCacheMiss(key)
+                            future
+                        }
+                    }
+            }
+
+        } else {
+            // not caching
+            future
+        }
+    }
+
+    protected def cacheUpdate(doc: W, key: Any, future: => Future[Winfo])(
+        implicit transid: TransactionId, logger: Logging, ec: ExecutionContext): Promise[Winfo] = {
+
+        val valPromise = Promise[W]
+        val promise = Promise[Winfo]
+
+        if (cacheEnabled) {
+            logger.info(this, s"caching $key")
+
+            val desiredEntry = Entry(transid, WriteInProgress, valPromise.future)
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) map {
+                // ... and see what we get back
+                case actualEntry =>
+                    if (transid == actualEntry.transid) {
+                        //
+                        // then we won the race, and now own the entry
+                        //
+                        logger.debug(this, s"Write initiated $key")
+                        listenForWriteDone(doc, key, actualEntry, future, valPromise, promise)
+
+                    } else {
+                        //
+                        // then we either lost the race, or there is an operation in progress on this key
+                        //
+                        logger.debug(this, s"Write under ${actualEntry.state.get} $key")
+                        promise completeWith future
+                    }
+            }
+
+        } else {
+            // not caching
+            promise completeWith future
+        }
+
+        promise
+    }
+
+    def getSize() = cache.size
```

Comment:
As per scala idiom: Use `def size = cache.size`. `()` indicate sideeffects, getters usually omit the `get`.